Move simple consumer code to separate project

omkreddy opened the pull request:
- Move old consumer related code to kafka-rest-scala-consumer subproject
- Move few common classes to kafka-rest-common subproject
- kafka-rest main project depends on kafka-rest-common, kafka-rest-scala-consumer
- publish shaded jar for kafka-rest-scala-consumer by relocating Kafka 1.1.0 classes

omkreddy commented:
[clabot:check]

omkreddy commented:
I moved old consumer code to a sub project and used shaded jar to avoid runtime errors.
Let me know your thoughts on this approach.

stanislavkozlovski commented:
Can we create a CPKAFKA JIRA better describing the motivation behind this change? This will ease other reviewers' work

ijuma commented on kafka-rest-scala-consumer/pom.xml:
```diff
@@ -0,0 +1,129 @@
+        </dependency>
+        <dependency>
+            <groupId>io.confluent</groupId>
+            <artifactId>kafka-avro-serializer</artifactId>
```
There are transitive dependencies to Kafka.

omkreddy replied:
SimpleConsumerManager class uses Decoder/KafkaAvroDecoder classes. Since we are relocating kafka.serializer.Decoder classes, we also need to relocate kafka-json-serializer classes.

ijuma commented:
Thanks for the PR @omkreddy and review @stanislavkozlovski.

> Something worthy of note is that these changes will now make our Producer and v2 consumer endpoints use Kafka's 2.3.0-SNAPSHOT codebase. I'm not sure of the implications of this but I personally find it a bit unsettling considering this will go out in minor version bump (CP 5.2->5.3)

This is fine since there are no compatibility implications for the REST API. I don't think we have any Java classes that are considered public API in this project. @ewencp to confirm.

omkreddy commented on kafka-rest-scala-consumer/src/main/java/io/confluent/kafkarest/SimpleConsumerConfig.java:
```diff
@@ -0,0 +1,292 @@
+
+import static io.confluent.common.config.ConfigDef.Range.atLeast;
+
+public class SimpleConsumerConfig extends RestConfig {
```
We can move KafkaRestConfig  to kafka-rest-common. I don't see any major downside. Initially wanted to minimize the common classes. Updated the PR.  Moved the KafkaRestConfig to kafka-common module.

stanislavkozlovski replied:
It's not necessarily bad to have common classes if we don't need to change them, right?

omkreddy commented on kafka-rest-scala-consumer/src/main/java/io/confluent/kafkarest/ScalaConsumersContext.java:
```diff
@@ -0,0 +1,77 @@
+  private MetadataObserver metadataObserver(final SimpleConsumerConfig config) {
+    if (StringUtil.isNotBlank(config.getString(ZOOKEEPER_CONNECT_CONFIG))) {
+      zkUtils = ZkUtils.apply(config.getString(ZOOKEEPER_CONNECT_CONFIG),
+          config.getInt(KAFKACLIENT_ZK_SESSION_TIMEOUT_MS_CONFIG),
```
I changed the defaults to old values. maybe we will make the timeout changes in separate PR.

stanislavkozlovski replied:
Probably a good idea 👍

omkreddy commented on kafka-rest-scala-consumer/pom.xml:
```diff
@@ -0,0 +1,129 @@
+                        </relocation>
+                        <relocation>
+                            <pattern>io.confluent.kafka.serializers</pattern>
+                            <shadedPattern>iio.confluent.kafka.shaded.serializers</shadedPattern>
```
Fixed the typo.

omkreddy commented on kafka-rest/src/main/java/io/confluent/kafkarest/extension/KafkaRestContextProvider.java:
```diff
@@ -15,85 +15,53 @@
-      } else if (mdObserver == null) {
-        mdObserver = new MetadataObserver(zkUtils);
+      if (scalaConsumersContext == null) {
+        SimpleConsumerConfig scalaConfog = null;
```
Removed the SimpleConsumerConfig class. Now we use KafkaRestConfig.

omkreddy commented:
@stanislavkozlovski Thanks for detailed review and thanks for testing the patch. Addresses the review comments. Pls take a look when you get a chance.

stanislavkozlovski commented on kafka-rest/src/test/java/io/confluent/kafkarest/unit/RootResourceTest.java:
```diff
@@ -42,7 +42,7 @@
 
   public RootResourceTest() throws RestConfigException {
-    ctx = new DefaultKafkaRestContext(config, null, null, null, null, null, null);
+    ctx = new DefaultKafkaRestContext(config, null, null, null, null );
```
nit: space after last null

omkreddy commented on kafka-rest/src/test/java/io/confluent/kafkarest/integration/AbstractProducerTest.java:
```diff
@@ -51,16 +51,16 @@
                                                int partition,
                                                List<? extends ProduceRecord<K, V>> records,
-                                               Decoder<K> keyDecoder, Decoder<K> valueDecoder,
+                                               String keySerializer, String valueSerializer,
```
changed the param names to keySerializerClassName/valueSerializerClassName for clarity.s

omkreddy commented on kafka-rest/src/test/java/io/confluent/kafkarest/integration/TestKafkaRestApplication.java:
```diff
@@ -28,48 +26,28 @@
  */
 public class TestKafkaRestApplication extends KafkaRestApplication {
```
We use TestKafkaRestApplication in ClusterTestHarness. We use it to inject test components in few tests like
https://github.com/omkreddy/kafka-rest-confluent/blob/refactor-scala-consumer/kafka-rest/src/test/java/io/confluent/kafkarest/integration/AbstractConsumerTest.java#L79

omkreddy commented:
@stanislavkozlovski Thanks for the review. Updated the PR.

omkreddy commented:
@stanislavkozlovski Thanks for the review. Merging to master. 
Apologies for anyone who has to fix conflicts in their PR.
